tests/test_spark.py

```diff
@@ -0,0 +1,123 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+
+import logging
+from time import sleep
+from typing import List
+
+import pytest
+from docker import DockerClient
+from docker.models.containers import Container
+from docker.types import Mount
+from granulate_utils.metrics import MetricsSnapshot
+from granulate_utils.metrics.metrics import (
+    SPARK_AGGREGATED_STAGE_METRICS,
+    SPARK_APPLICATION_DIFF_METRICS,
+    SPARK_APPLICATION_GAUGE_METRICS,
+    SPARK_EXECUTORS_METRICS,
+)
+from granulate_utils.metrics.sampler import BigDataSampler
+from pytest import LogCaptureFixture
+
+from gprofiler.log import get_logger_adapter
+from tests.conftest import _build_image
+from tests.utils import _wait_container_to_start
+
+# `BigDataSampler` receives a logger as an argument.
+logger = get_logger_adapter("gprofiler_spark_test")
+
+# List that includes all the metrics that are expected to be collected by `BigDataSampler` in SA mode.
+EXPECTED_SA_METRICS_KEYS = [
+    metric
+    for metrics_dict in (
+        SPARK_APPLICATION_GAUGE_METRICS,
+        SPARK_APPLICATION_DIFF_METRICS,
+        SPARK_AGGREGATED_STAGE_METRICS,
+        SPARK_EXECUTORS_METRICS,
+    )
+    for metric in metrics_dict.values()
+]
+
+SPARK_MASTER_HOST = "127.0.0.1"
+
+DISCOVER_INTERVAL_SECS = 5
+DISCOVER_TIMEOUT_SECS = 60
+
+
+@pytest.fixture(scope="function")
+def sparkpi_container(docker_client: DockerClient, application_docker_mounts: List[Mount]) -> Container:
+    """
+    This fixture is responsible for running SparkPi application in a container.
+    See `containers/spark/Dockerfile`
+    """
+    spark_image = _build_image(docker_client=docker_client, runtime="spark")
```

Comment:
I see that _build_image passes rm=True to docker build, meaning cache is not retained. So this rebuilds the image for every test. The image should be a separate fixture with wider scope.

Reply:
seems like a broader issue - `build_image` is used by `sparkpi_container` and `application_docker_image`, both are function scoped..
WDYT @Jongy

Reply:
It's not what `rm` means:
> rm (bool) – Remove intermediate containers. The docker build command now defaults to --rm=true, but we have kept the old default of False to preserve backward compatibility

Not intermediate images but intermediate containers, we do want to remove them